contracts/BonusRewards.sol

```diff
@@ -3,11 +3,361 @@
+      accRewardsPerToken: 0,
+      remBonus: received
+    }));
+  }
+
+  /// @notice called by authorizers only, update weeklyRewards (if not ended), or update startTime (only if rewards not started, 0 is ignored)
+  function updateBonus(
+    address _lpToken,
+    address _bonusTokenAddr,
+    uint256 _weeklyRewards,
+    uint48 _startTime
+  ) external override nonReentrant notPaused {
+    require(_isAuthorized(msg.sender, allowedTokenAuthorizers[_lpToken][_bonusTokenAddr]), "BonusRewards: not authorized caller");
+    require(_startTime == 0 || _startTime > block.timestamp, "BonusRewards: startTime in the past");
+
+    // make sure the pool is in the right state (exist with no active bonus at the moment) to add new bonus tokens
+    Pool memory pool = pools[_lpToken];
+    require(pool.lastUpdatedAt > 0, "BonusRewards: pool does not exist");
+    Bonus[] memory bonuses = pool.bonuses;
+    for (uint256 i = 0; i < bonuses.length; i++) {
+      if (bonuses[i].bonusTokenAddr == _bonusTokenAddr && bonuses[i].endTime > block.timestamp) {
+        Bonus storage bonus = pools[_lpToken].bonuses[i];
+        _updatePool(_lpToken); // update pool with old weeklyReward to this block
+        if (bonus.startTime >= block.timestamp) {
+          // only honor new start time, if program has not started
+          if (_startTime >= block.timestamp) {
+            bonus.startTime = _startTime;
+          }
+          bonus.endTime = uint48(bonus.remBonus * WEEK / _weeklyRewards + bonus.startTime);
+        } else {
+          uint256 remBonusToDistribute = (bonus.endTime - block.timestamp) * bonus.weeklyRewards / WEEK;
+          bonus.endTime = uint48(remBonusToDistribute * WEEK / _weeklyRewards + block.timestamp);
+        }
+        bonus.weeklyRewards = _weeklyRewards;
+      }
+    }
+  }
+
+  /// @notice extend the current bonus program, the program has to be active (endTime is in the future)
+  function extendBonus(
+    address _lpToken,
+    uint256 _poolBonusId,
+    address _bonusTokenAddr,
+    uint256 _transferAmount
+  ) external override nonReentrant notPaused {
+    require(_isAuthorized(msg.sender, allowedTokenAuthorizers[_lpToken][_bonusTokenAddr]), "BonusRewards: not authorized caller");
+
+    Bonus memory bonus = pools[_lpToken].bonuses[_poolBonusId];
+    require(bonus.bonusTokenAddr == _bonusTokenAddr, "BonusRewards: bonus and id dont match");
+    require(bonus.endTime > block.timestamp, "BonusRewards: bonus program ended, please start a new one");
+
+    IERC20 bonusTokenAddr = IERC20(_bonusTokenAddr);
+    uint256 balanceBefore = bonusTokenAddr.balanceOf(address(this));
+    bonusTokenAddr.safeTransferFrom(msg.sender, address(this), _transferAmount);
+    uint256 received = bonusTokenAddr.balanceOf(address(this)) - balanceBefore;
+    // endTime is based on how much tokens transfered v.s. planned weekly rewards
+    uint48 endTime = uint48(received * WEEK / bonus.weeklyRewards + bonus.endTime);
```

Comment:
Shouldn't this be aware of the previous balance to determine the right end time? It's only using the received balance to determine the endtime no? When there are tokens that are already in this contract?

Reply:
`Shouldn't this be aware of the previous balance to determine the right end time?`
No, cause the week rewards determines how long to extend. It is extending on top of the previous `endTime`.